src/main/java/seedu/mtracker/filemanager/InstrumentDecoder.java

```diff
@@ -0,0 +1,74 @@
+package seedu.mtracker.filemanager;
+
+import seedu.mtracker.instrument.Instrument;
+import seedu.mtracker.instrument.InstrumentManager;
+import seedu.mtracker.instrument.subinstrument.Crypto;
+import seedu.mtracker.instrument.subinstrument.Etf;
+import seedu.mtracker.instrument.subinstrument.Forex;
+import seedu.mtracker.instrument.subinstrument.Stock;
+
+import java.util.List;
+
+public class InstrumentDecoder {
+    public static void readFile(InstrumentManager instrumentManager, List<String> fileData) {
+        fileData.stream()
+                .forEach((line) -> {
+                    String[] textSegment = line.split(";", -1);
+                    switch (textSegment[0]) {
+                    case ("Crypto"):
+                        addCryptoToList(textSegment, instrumentManager);
+                        break;
+                    case ("Stock"):
+                        addStockToList(textSegment, instrumentManager);
+                        break;
+                    case ("Forex"):
+                        addForexToList(textSegment, instrumentManager);
+                        break;
+                    default:
+                        addEtfToList(textSegment, instrumentManager);
+                    }
+                });
+    }
+
+    public static void addCryptoToList(String[] textSegment, InstrumentManager instrumentManager) {
+        String name = textSegment[1];
+        double currentPrice = Double.parseDouble(textSegment[2]);
+        String sentiment = textSegment[3];
+        String expiry = textSegment[4];
+        String remarks = textSegment[5];
+        Instrument crypto = new Crypto(name, currentPrice, sentiment, expiry, remarks);
+        instrumentManager.addInstrument(crypto);
```

Comment:
define the magic indexes. To error handle, we should validate each field, like what we did for addParser, so if possible check out whether can make file not writable. In anycase for error handling can do in separate PR. I have plans to refactor the validations out of parser -> https://github.com/AY2122S1-CS2113T-T12-1/tp/issues/92
Once refactored along with standardising the custom exceptions, we can work on error handling for file

Reply:
I noticed that other instrument type remarks require a different index, if so just define specifically REMARK_FOREX_INDEX for example for the different types

Reply:
added one custom exception for now can handle the rest in a separate pr